esp-hal/src/lib.rs

```diff
@@ -494,68 +494,92 @@ pub struct Config {
     pub psram: psram::PsramConfig,
 }
 
-/// Initialize the system.
-///
-/// This function sets up the CPU clock and watchdog, then, returns the
-/// peripherals and clocks.
-pub fn init(config: Config) -> Peripherals {
-    system::disable_peripherals();
-
-    let mut peripherals = Peripherals::take();
-
-    // RTC domain must be enabled before we try to disable
-    let mut rtc = crate::rtc_cntl::Rtc::new(&mut peripherals.LPWR);
-
-    #[cfg(not(any(esp32, esp32s2)))]
-    if config.watchdog.swd {
-        rtc.swd.enable();
-    } else {
-        rtc.swd.disable();
-    }
-
-    match config.watchdog.rwdt {
-        WatchdogStatus::Enabled(duration) => {
-            rtc.rwdt.enable();
-            rtc.rwdt
-                .set_timeout(crate::rtc_cntl::RwdtStage::Stage0, duration);
-        }
-        WatchdogStatus::Disabled => {
-            rtc.rwdt.disable();
+fn init_internal<'a>(peripherals: Option<&'a mut Peripherals>, config: Config) {
+    if let Some(peripherals) = peripherals {
```

Comment:
I'd prefer not passing `Option<&mut Peripherals>` - you can (and you do) check outside and only call this when needed.

Reply:
Just to clarify -- you're suggesting (only that) I should have `try_init` return nothing?

Reply:
This is what I meant:

```rust
/// Initialize the system.
///
/// This function sets up the CPU clock and watchdog, then, returns the
/// peripherals and clocks.
pub fn init(config: Config) -> Peripherals {
    system::disable_peripherals();

    let mut peripherals = if let Some(p) = Peripherals::take() {
        p
    } else {
        panic!("init called more than once!");
    };

    init_internal(&mut peripherals, config);

    peripherals
}

/// Initializes the system only if necessary, then returns the peripherals.
///
/// If [`init`] has already been executed, simply steals and returns the
/// peripherals. (This check uses [`AtomicBool`](core::sync::atomic::AtomicBool)
/// to be as thread-safe as possible)
///
/// Otherwise, the initialization process is executed (and then the peripherals
/// are stolen and handed off to you).
pub unsafe fn try_init(config: Config) -> Peripherals {
    if let Some(mut peripherals) = Peripherals::take() {
        init_internal(&mut peripherals, config);
    }
    Peripherals::steal()
}
```